File: Lecture6.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm

x_range = np.linspace(2, 10, 1000)  # 设置 x 的范围为 [2,10]
y_norm = norm.pdf(x_range, 6.25, 0.75)

# 对y_norm进行采样，得到1000个样本值
samples = np.random.normal(6.25, 0.75, 1000)

# 显示前 10 个样本值
first_10_samples = samples[:10]
print("前 10 个样本值：", first_10_samples)

#  绘制图像
plt.figure(figsize=(10, 6))
plt.plot(x_range, norm.pdf(x_range, 6.25, 0.75), linewidth=2, color='r', label='Normal Distribution PDF')
count, bins, ignored = plt.hist(samples, bins=30, density=True, alpha=0.6, color='grey', edgecolor='black',
                                label='Sampled Histogram')

plt.legend()
plt.show()

import numpy as np
import scipy.stats as st
import pandas as pd
import seaborn as sns

np.random.seed(2024)

current = 3  # 假设theta(n)为3

proposal = st.norm(current, 1).rvs()  # 从当前正态分布中抽出一个样本

print("从建议分布中新采样 θ(n+1)为：", proposal)

# 设置先验
prior = st.norm(loc=3, scale=1)


def likelihood(theta):
    Y = 6  # 假设数据 Y 为 6
    return st.norm(loc=theta, scale=0.75).pdf(Y)  # 注意：这里为方便演示固定 scale = 0.75


# 计算建议位置(n+1)的未归一化的后验概率值（先验*似然）
proposal_posterior = prior.pdf(proposal) * likelihood(proposal)

# 计算当前位置(n)的未归一化的后验概率值（先验*似然）
current_posterior = prior.pdf(current) * likelihood(current)

# 计算接受概率α，为两者概率值之比
alpha = min(1, proposal_posterior / current_posterior)

# 打印出接受概率α
print("后验比为：", proposal_posterior / current_posterior, ", alpha为:", alpha)

# 根据接受概率α进行抽样，抽样内容为建议位置和当前位置
next_stop = np.random.choice([proposal, current], 1, p=[alpha, 1 - alpha])

# 打印出下一个位置的值
next_stop[0]


# 从第一段代码我们可以看到此时的接受概率α=1，因此接受了建议值作为我们的下一个值

def one_mh_iteration(current, sigma=1):
    """
    def后面为函数值，current为输入值，作为建议分布(正态分布)的均值

    接下来的代码和之前一样

    return 则是该函数返回的值，我们将建议值，接受概率，和下一个位置这三个值组成了一个数据框进行返回
    """
    proposal = st.norm(current, sigma).rvs()

    prior = st.norm(loc=3, scale=1)

    def likelihood(theta):
        # 假设数据 Y 为 6
        Y = 6
        return st.norm(loc=theta, scale=0.75).pdf(Y)

    proposal_posterior = prior.pdf(proposal) * likelihood(proposal)
    current_posterior = prior.pdf(current) * likelihood(current)
    alpha = min(1, proposal_posterior / current_posterior)
    next_stop = np.random.choice([proposal, current], 1, p=[alpha, 1 - alpha])
    return pd.DataFrame({"proposal": [proposal],
                         "alpha": [alpha],
                         "next_stop": [next_stop[0]]})


np.random.seed(8)
one_mh_iteration(current=3)

# 变换不同的随机数种子，其实也是生成不同的建议值
np.random.seed(83)
one_mh_iteration(current=3)


def mh_tour(N, sigma=1):
    """
    N为迭代次数，sigma为正态建议分布的标准差

    我们在单次采样函数的基础上叠加了一个循环
    将每次的采样结果存在mu[i]中，
    在每次采样结束后，将采样结果替换为当前位置

    返回值为迭代次数，和每次采样得到的结果
    """
    current = 3
    mu = np.zeros(N)

    for i in range(N):
        sim = one_mh_iteration(current, sigma)
        mu[i] = sim["next_stop"][0]
        current = sim["next_stop"][0]

    return pd.DataFrame({"iteration": range(1, N + 1),
                         'mu': mu})


# 调用定义好的函数，将采样次数设为5000
np.random.seed(84735)
mh_simulation = mh_tour(N=5000)
mh_simulation.tail()

import matplotlib.pyplot as plt

# 生成一行两列的画布
fig, axs = plt.subplots(1, 2, figsize=(20, 5))

# density plot：在第一列绘制出采样结果的分布
axs[0].hist(mh_simulation["mu"],
            edgecolor="white",
            color="grey",
            alpha=0.7,
            bins=20,
            density=True)
axs[0].set_xlabel("mu", fontsize=16)
axs[0].set_ylabel("density", fontsize=16)

# 绘制分布外围线条
x_norm = np.linspace(2, 10, 10000)
y_norm = st.norm.pdf(x_norm, loc=mh_simulation["mu"].mean(), scale=mh_simulation["mu"].std())

axs[0].plot(x_norm, y_norm, color='blue')

# trace plot：在第二列绘制出每一次的采样结果
axs[1].plot(mh_simulation["iteration"], mh_simulation["mu"],
            color="grey", )
axs[1].set_xlabel("iteration", fontsize=16)
axs[1].set_ylabel("mu", fontsize=16)

import arviz as az
import matplotlib.pyplot as plt

ax = az.plot_trace({"mu": mh_simulation["mu"]})

ax[0, 0].set_xlim(2, 10)

import numpy as np
import matplotlib.pyplot as plt
import arviz as az
import scipy.stats as st
import pandas as pd


def one_mh_iteration(current, sigma=1):
    """
    def后面为函数值，current为输入值，作为建议分布(正态分布)的均值

    接下来的代码和之前一样

    return 则是该函数返回的值，我们将建议值，接受概率，和下一个位置这三个值组成了一个数据框进行返回
    """
    proposal = st.norm(current, sigma).rvs()

    prior = st.norm(loc=3, scale=1)

    def likelihood(theta):
        # 假设数据 Y 为 6
        Y = 6
        return st.norm(loc=theta, scale=0.75).pdf(Y)

    proposal_posterior = prior.pdf(proposal) * likelihood(proposal)
    current_posterior = prior.pdf(current) * likelihood(current)
    alpha = min(1, proposal_posterior / current_posterior)
    next_stop = np.random.choice([proposal, current], 1, p=[alpha, 1 - alpha])
    return pd.DataFrame({"proposal": [proposal],
                         "alpha": [alpha],
                         "next_stop": [next_stop[0]]})


def mh_tour(N, sigma=1):
    """
    N为迭代次数，sigma为正态建议分布的标准差

    我们在单次采样函数的基础上叠加了一个循环
    将每次的采样结果存在mu[i]中，
    在每次采样结束后，将采样结果替换为当前位置

    返回值为迭代次数，和每次采样得到的结果
    """
    current = 3
    mu = np.zeros(N)

    for i in range(N):
        sim = one_mh_iteration(current, sigma)
        mu[i] = sim["next_stop"][0]
        current = sim["next_stop"][0]

    return pd.DataFrame({"iteration": range(1, N + 1),
                         'mu': mu})


# ===========================================================================
#                            请修改 ... 中的值。
# ===========================================================================
np.random.seed(84735)

mh_simulation = mh_tour(N=5000, sigma=...)
az.plot_trace({"mu": mh_simulation["mu"]})

# ===========================================================================
#                            可以自行复制代码多试几次
# ===========================================================================
mh_simulation = mh_tour(N=5000, sigma=...)
az.plot_trace({"mu": mh_simulation["mu"]})

##PyMC实战
import pymc as pm
import arviz as az

# 定义观测值Y=9
Y = 9

# 创建一个 PyMC 模型名为 bb_model
bb_model = pm.Model()

# 使用with语句将模型与变量联系起来
with bb_model:
    # 定义一个符合beta先验分布的变量pi
    pi = pm.Beta('pi', alpha=2, beta=2)
    # 定义符合binomial分布的变量likelihood，其中总次数为10，成功概率为pi，观测值为Y
    likelihood = pm.Binomial('likelihood', n=10, p=pi, observed=Y)

pm.model_to_graphviz(bb_model)

# 采样过程仍在该容器中进行
with bb_model:
    trace = pm.sample(draws=5000,  # 使用mcmc方法进行采样，draws为采样次数
                      chains=1,  # 设置MCMC的链数
                      random_seed=84735)  # 设置随机状态，以获得与notebook相同的结果

# 可视化采样结果
az.plot_trace(trace)

# 选取第一条Makov链的前20个采样结果和前200个结果
# 从trace中提取pi的值，使用sel(chain=0)表示为选择第一条markov链。使用.values[:20]表示选取前20个采样结果
samples_20 = trace.posterior["pi"].sel(chain=0).values[:20]
samples_200 = trace.posterior["pi"].sel(chain=0).values[:200]

# 绘图
# az.plot_trace表示绘制前20个采样结果的轨迹图，其中pi为变量名，samples_20为采样结果
az.plot_trace({"pi": samples_20})
az.plot_trace({"pi": samples_200})

import pandas as pd

post_pi = pd.DataFrame({"pi": trace.posterior["pi"].values.reshape(-1)})
post_pi

import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import scipy.stats as st

fig, (axs1, axs2) = plt.subplots(1, 2, figsize=(15, 4))

# 绘制采样结果直方图
sns.histplot(data=post_pi,
             x="pi",
             bins=30,
             ax=axs1,
             edgecolor='#20699d',
             color="#6497b1",
             alpha=1)

# 绘制采样结果密度分布图
sns.kdeplot(data=post_pi,
            x="pi",
            color='#6497b1',
            fill=True,
            alpha=1,
            ax=axs2)

# 绘制真实后验分布图
x = np.linspace(0.2, 1, 10000)
# 真实的后验分布 beta（alpha + y, beta + n - y）
y = st.beta.pdf(x, 11, 3)
axs2.plot(x, y, color='black')

sns.despine()

##练习
import pymc as pm
import arviz as az

# ===========================================================================
#                            请修改 ... 中的值。
# ===========================================================================

observed_data = [...]

# 1. 设立容器

with pm.Model() as nn_model:
    # 2. 定义先验
    mu_prior = pm.Normal('...', mu=..., sigma=...)
    # 3. 定义似然
    Y_obs = pm.Normal('...', mu=..., sigma=..., observed=...)
# ===========================================================================
#                            请修改 ... 中的值。
# ===========================================================================
with nn_model:
    trace = pm.sample(draws=...,  # 例如采样次数设为500-5000次
                      chains=1,  # 链为1条
                      random_seed=84735)
trace

az.plot_trace(trace)

# 从 MCMC 采样结果中提取参数 mu 的后验分布样本
post_mu = pd.DataFrame({"mu": trace.posterior["mu_prior"].values.reshape(-1)})

# ===========================================================================
#                            为了和真实的后验分布进行比较，计算真实的后验分布。
# ===========================================================================


# 先验分布的均值和标准差
mu_prior = 300  # 先验均值
sigma_prior = 50  # 先验标准差

# 观测数据的标准差 (已知)
sigma_obs = 20

# 计算观测数据的数量和均值
n = len(observed_data)
y_mean = np.mean(observed_data)

# 计算后验分布的均值和方差
posterior_mean = (sigma_obs ** 2 * mu_prior + n * sigma_prior ** 2 * y_mean) / (n * sigma_prior ** 2 + sigma_obs ** 2)
posterior_variance = (sigma_prior ** 2 * sigma_obs ** 2) / (n * sigma_prior ** 2 + sigma_obs ** 2)
posterior_std = np.sqrt(posterior_variance)

print(f"后验分布的均值: {posterior_mean}")
print(f"后验分布的标准差: {posterior_std}")

fig, (axs1, axs2) = plt.subplots(1, 2, figsize=(15, 4))

sns.histplot(data=post_mu,
             x="mu",
             bins=30,
             ax=axs1,
             edgecolor='#20699d',
             color="#6497b1",
             alpha=1)

sns.kdeplot(data=post_mu,
            x="mu",
            color='#6497b1',
            fill=True,
            alpha=1,
            ax=axs2)

# 计算真实的后验分布
x = np.linspace(200, 400, 10000)
y = st.norm.pdf(x, posterior_mean, posterior_std)
axs2.plot(x, y, color='black')

sns.despine()

File: Lecture6.R
# 安装和加载包
options(repos = c(CRAN = "https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
if (!requireNamespace('pacman', quietly = TRUE)) {
  install.packages('pacman')
}
pacman::p_load("tidyverse","ggplot2", "dplyr","gridExtra","papaja", "patchwork","bayesplot","rstan")
options(warn = -1)  # 抑制警告
# 设置 x 的范围为 [2, 10]
x_range <- seq(2, 10, length.out = 1000)

# 计算正态分布的概率密度函数
y_norm <- dnorm(x_range, mean = 6.25, sd = 0.75)

# 对正态分布进行采样，得到1000个样本值
samples <- rnorm(1000, mean = 6.25, sd = 0.75)

# 显示前 10 个样本值
cat("前 10 个样本值：", head(samples, 10))

# 绘制图像
options(repr.plot.width=10, repr.plot.height=6) 
df <- data.frame(x = x_range, y = y_norm)
ggplot2::ggplot(df, aes(x = x, y = y)) +
  ggplot2::geom_histogram(aes(x = samples, y = ..density..),  bins = 45, fill = "lightgrey", color = "black") +
  ggplot2::geom_line(color = 'red', size = 1.5)+ 
  ggplot2::scale_y_continuous(expand = c(0, 0) , limits = c(0, 0.55)) +
  papaja::theme_apa()

set.seed(2024)

current <- 3                             # 假设theta^n为3

proposal <- rnorm(1, mean = current, sd = 1)    # 从当前正态分布中抽出一个样本

cat("从建议分布中新采样 θ(n+1)为：",proposal)
# 设置先验
prior <- dnorm(proposal, mean = 3, sd = 1)  # 先验概率密度

# 定义似然函数
likelihood <- function(theta) {
  Y <- 6  # 假设数据 Y 为 6
  return(dnorm(Y, mean = theta, sd = 0.75))  # 固定 sd=0.75
}

# 计算建议位置的未归一化的后验概率值（先验 * 似然）
proposal_posterior <- prior * likelihood(proposal)

# 计算当前位置的未归一化的后验概率值（先验 * 似然）
current_posterior <- prior * likelihood(current)

# 计算接受概率α，为两者概率值之比
alpha <- min(1, proposal_posterior / current_posterior)

# 打印出接受概率α
cat("后验比为：", proposal_posterior / current_posterior, ", alpha为:", alpha, "\n")

# 根据接受概率α进行抽样，抽样内容为建议位置和当前位置
next_stop <- sample(c(proposal, current), 1, prob = c(alpha, 1 - alpha))

# 打印出下一个位置的值
cat("下一个位置的值为:", next_stop, "\n")
# 从第一段代码我们可以看到此时的接受概率α=1，因此接受了建议值作为我们的下一个值

one_mh_iteration <- function(current, sigma = 1) {
  
  # 提议值:从均值为current，方差为sigma（默认值为1）的正态分布中抽样
  proposal <- rnorm(1, mean = current, sd = sigma)
  
  # 设置先验并计算概率密度
  prior <- dnorm(proposal, mean = 3, sd = 1)  
  
  # 定义似然函数
  likelihood <- function(theta) {
    Y <- 6  # 假设数据 Y 为 6
    return(dnorm(Y, mean = theta, sd = 0.75))  # 计算似然
  }
  
  # 计算未归一化的后验概率
  proposal_posterior <- prior * likelihood(proposal)
  current_posterior <- prior * likelihood(current)
  
  # 计算接受概率 alpha
  if (is.na(proposal_posterior) || is.na(current_posterior) || current_posterior <= 0) {
    alpha <- 0
  } else {
    alpha <- min(1, proposal_posterior / current_posterior)
  }
  
  # 根据接受概率 alpha 进行抽样
  next_stop <- sample(c(proposal, current), 1, prob = c(alpha, 1 - alpha), replace = TRUE)
  
  # 返回建议值、接受概率和下一个位置组成的数据框
  result <- data.frame(proposal = proposal, alpha = alpha, next_stop = next_stop)
  
  return(result)
}

# 设置随机种子
set.seed(2024)

# 调用函数并显示结果
results <- one_mh_iteration(current = 3)
print(results, row.names = FALSE) 

# 变换不同的随机数种子，其实也是生成不同的建议值
set.seed(83)

# 调用函数并显示结果
results <- one_mh_iteration(current = 3)
print(round(results, digits = 4), row.names = FALSE) #用round对数据框内数字四舍五入，保留4位小数

mh_tour <- function(N, sigma = 1) {
  current <- 3
  mu <- numeric(N)  # 创建一个长度为 N 的零向量
  
  # 循环进行 N 次迭代
  for (i in 1:N) {
    sim <- one_mh_iteration(current, sigma)  # 调用采样函数
    mu[i] <- sim$next_stop  # 保存当前的下一个位置
    current <- sim$next_stop  # 更新当前值
  }
  
  # 返回包含迭代次数和每次采样结果的数据框
  result <- data.frame(iteration = 1:N,mu = mu)
  
  return(result)
}

# 调用定义好的函数，将采样次数设为5000
set.seed(84735)
mh_simulation <- mh_tour(N=5000)
tail(mh_simulation, n = 5)

# 绘制密度图
density_plot <- ggplot2::ggplot(mh_simulation, aes(x = mu)) +
  ggplot2::geom_histogram(aes(y = ..density..), 
                          bins = 30, 
                          color = "white", 
                          fill = "darkgrey", 
                          alpha = 0.7) +
  ggplot2::geom_density(aes(y = ..density..), 
                        color = "#6497b1", 
                        size = 1) +
  ggplot2::labs(x = "mu", y = "density") +
  papaja::theme_apa()  + 
  ggplot2::scale_y_continuous(expand = c(0, 0) , limits = c(0, 0.7)) 

# 绘制轨迹图
trace_plot <- ggplot2::ggplot(mh_simulation, aes(x = iteration, y = mu)) +
  ggplot2::geom_line(color = "#6497b1") +
  ggplot2::labs(x = "iteration", y = "mu") +
  papaja::theme_apa()  + 
  ggplot2::scale_y_continuous(limits = c(3, 9)) 

# 将两个图并排显示
options(repr.plot.width=16, repr.plot.height=7) 
density_plot + trace_plot

#创建绘图数据框
sample <- as.data.frame(mh_simulation$mu)

#采样密度图
dens <- bayesplot::mcmc_dens(sample) +    
  ggplot2::labs(x = "mu", y = "density") +
  papaja::theme_apa()

#采样轨迹图
trace <- bayesplot::mcmc_trace(sample) +  
  ggplot2::labs(x = "iteration", y = "density") +
  papaja::theme_apa()

dens + trace

one_mh_iteration <- function(current, sigma = 1) {
  
  # 提议值:从均值为current，方差为sigma（默认值为1）的正态分布中抽样
  proposal <- rnorm(1, mean = current, sd = sigma)
  
  # 设置先验并计算概率密度
  prior <- dnorm(proposal, mean = 3, sd = 1)  
  
  # 定义似然函数
  likelihood <- function(theta) {
    Y <- 6  # 假设数据 Y 为 6
    return(dnorm(Y, mean = theta, sd = 0.75))  # 计算似然
  }
  
  # 计算未归一化的后验概率
  proposal_posterior <- prior * likelihood(proposal)
  current_posterior <- prior * likelihood(current)
  
  # 计算接受概率 alpha
  if (is.na(proposal_posterior) || is.na(current_posterior) || current_posterior <= 0) {
    alpha <- 0
  } else {
    alpha <- min(1, proposal_posterior / current_posterior)
  }
  
  # 根据接受概率 alpha 进行抽样
  next_stop <- sample(c(proposal, current), 1, prob = c(alpha, 1 - alpha), replace = TRUE)
  
  # 返回建议值、接受概率和下一个位置组成的数据框
  result <- data.frame(proposal = proposal, alpha = alpha, next_stop = next_stop)
  
  return(result)
}

mh_tour <- function(N, sigma = 1) {
  current <- 3
  mu <- numeric(N)  # 创建一个长度为 N 的零向量
  
  # 循环进行 N 次迭代
  for (i in 1:N) {
    sim <- one_mh_iteration(current, sigma)  # 调用采样函数
    mu[i] <- sim$next_stop  # 保存当前的下一个位置
    current <- sim$next_stop  # 更新当前值
  }
  
  # 返回包含迭代次数和每次采样结果的数据框
  result <- data.frame(iteration = 1:N,mu = mu)
  
  return(result)
}

#===========================================================================
#                            请修改 sigma 的值。
#===========================================================================
set.seed(84735)
mh_simulation = mh_tour(N=5000, sigma=1)
#创建绘图数据框
sample_x <- as.data.frame(mh_simulation$mu)

#采样密度图
dens_x <- bayesplot::mcmc_dens(sample_x) +    
  ggplot2::labs(x = "mu", y = "density") +
  papaja::theme_apa()

#采样轨迹图
trace_x <- bayesplot::mcmc_trace(sample_x) +  
  ggplot2::labs(x = "iteration", y = "density") +
  ggplot2::scale_y_continuous(limits = c(2.5, 9)) +
  papaja::theme_apa()

dens_x + trace_x

##Stan实战
# 准备数据
stan_data <- list(n = 10,Y = 9)
# 使用Stan语法建构模型
stan_model_code <- "
data {  //数据块
  int<lower=0> n;       // 试验次数
  int<lower=0, upper=n> Y;  // 成功次数
}

parameters {  //参数块
  real<lower=0, upper=1> pi;    // 成功概率参数
}

model {  //模型块
  // 设置先验
  pi ~ beta(2, 2);
  
  // 设置似然
  Y ~ binomial(n, pi);
}
"
# 拟合模型
trace <- rstan::stan(
  model_code = stan_model_code,      # 定义的模型或模型文件路径
  data = stan_data,                  # 输入数据
  chains = 1,                        # 马尔可夫链数量
  iter = 5000,                       # 总迭代次数（每个链）
  warmup = 0,                        # 热身迭代次数（不保存）
  seed = 202409
)

#创建绘图数据框
idata <- rstan::extract(trace)
sample_pi <- as.data.frame(idata$pi)
#采样密度图
dens_pi <- bayesplot::mcmc_dens(sample_pi) +    
  ggplot2::labs(x = "pi", y = "density") +
  papaja::theme_apa()
# 绘制轨迹图
trace_pi <- rstan::traceplot(trace,color = '#6497b1')

dens_pi + trace_pi

# 选取第一条Makov链的前20个采样结果和前200个结果
sample_pi_20 <- as.data.frame(sample_pi[1:20, ])
sample_pi_200 <- as.data.frame(sample_pi[1:200, ])

# 绘图
options(repr.plot.width=16, repr.plot.height=5) 
#前20
dens_20 <- bayesplot::mcmc_dens(sample_pi_20) +    
  ggplot2::labs(x = "pi", y = "density") +
  papaja::theme_apa()
trace_20 <- bayesplot::mcmc_trace(sample_pi_20) +  
  ggplot2::labs(x = "pi", y = NULL) +
  papaja::theme_apa() 
dens_20 + trace_20

#前200
dens_200 <- bayesplot::mcmc_dens(sample_pi_200) +    
  ggplot2::labs(x = "pi", y = "density") +
  papaja::theme_apa()
trace_200 <- bayesplot::mcmc_trace(sample_pi_200) +  
  ggplot2::labs(x = "pi", y = NULL) +
  papaja::theme_apa() 
dens_200 + trace_200

rbind(head(sample_pi, 5), tail(sample_pi, 5))

x <- seq(0.2, 1, length.out = 10000)
# 真实的后验分布 beta（alpha + y, beta + n - y）
y <- dbeta(x, 11, 3)
posterior_data <- data.frame(x = x, y = y)

#绘制采样结果直方图
hist_pi <- bayesplot::mcmc_hist(sample_pi, bins=30) +    
  ggplot2::labs(x = "pi", y = "count") +
  papaja::theme_apa()

#绘制采样结果分布图和真实后验分布图
dens_pi_mix <- bayesplot::mcmc_dens(sample_pi) +  #采样结果
  ggplot2::geom_line(data = posterior_data, aes(x = x, y = y), color = "black",size = 0.8) +  #真实后验
  ggplot2::labs(x = "pi", y = "density") +
  papaja::theme_apa()

hist_pi + dens_pi_mix 

##练习（完整代码）
#===========================================================================
#                            请修改 ... 中的值。
#===========================================================================

#准备数据
data_list <- list(
  Y = c(320, 310, 280, 340, 300), # 观测数据
  n = 5                          # 观测数据的数量
)

#构建模型

stan_code <- "
data {
  int<lower=0> n;           // 观测数据的数量
  vector[n] Y;              // 观测数据
}

parameters {
  real mu;                  // 反应时间均值
}

model {
  // 先验分布
  mu ~ normal(300, 50);

  // 似然函数
  Y ~ normal(mu, 20);
}
"
#===========================================================================
#                            请修改 ... 中的值。
#===========================================================================
trace <- rstan::stan(
  model_code = stan_code,            # 定义的模型或模型文件路径
  data = data_list,                  # 输入数据
  chains = 1,                        # 马尔可夫链数量
  iter = 5000,                       # 总迭代次数
  warmup = 500,                        # 热身迭代次数
  seed = 202409
)

#创建绘图数据框
idata <- rstan::extract(trace)
sample_mu <- as.data.frame(idata$mu)
#采样密度图
dens_mu <- bayesplot::mcmc_dens(sample_mu) +    
  ggplot2::labs(x = "mu_prior", y = "density") +
  papaja::theme_apa()
# 绘制轨迹图
trace_mu <- rstan::traceplot(trace,color = '#6497b1')

dens_mu + trace_mu

#===========================================================================
#                            为了和真实的后验分布进行比较，计算真实的后验分布。
#===========================================================================


# 先验分布的均值和标准差
mu_prior <- 300  # 先验均值
sigma_prior <- 50  # 先验标准差

# 观测数据的标准差 (已知)
sigma_obs <- 20

# 假设观测数据
observed_data <- c(320, 310, 280, 340, 300)  # 替换为实际观测数据

# 计算观测数据的数量和均值
n <- length(observed_data)  
y_mean <- mean(observed_data)  

# 计算后验分布的均值和方差
posterior_mean <- (sigma_obs^2 * mu_prior + n * sigma_prior^2 * y_mean) / (n * sigma_prior^2 + sigma_obs^2)
posterior_variance <- (sigma_prior^2 * sigma_obs^2) / (n * sigma_prior^2 + sigma_obs^2)
posterior_std <- sqrt(posterior_variance)

# 输出结果
cat("后验分布的均值:", posterior_mean)
cat("后验分布的标准差:", posterior_std)

# 计算真实的后验分布
x <- seq(200, 400, length.out = 5000)  # 创建从200到400的序列
y <- dnorm(x, mean = posterior_mean, sd = posterior_std)  # 计算后验分布的概率密度
posterior_real <- data.frame(x = x, y = y)

#绘制采样结果直方图
hist_mu <- bayesplot::mcmc_hist(sample_mu, bins=30) +    
  ggplot2::labs(x = "mu", y = "count") +
  papaja::theme_apa()

#绘制采样结果分布图和真实后验分布图
dens_mu_mix <- bayesplot::mcmc_dens(sample_mu) +  #采样结果
  ggplot2::geom_line(data = posterior_real, aes(x = x, y = y), color = "black",size = 0.8) +  #真实后验
  ggplot2::labs(x = "mu", y = "density") +
  papaja::theme_apa()

options(repr.plot.width=16, repr.plot.height=6) 
hist_mu + dens_mu_mix 